lib/ipam/ipam.go

```diff
@@ -1383,6 +1344,62 @@ func (c ipamClient) convertBackendToIPAMConfig(cfg *model.IPAMConfig) *IPAMConfi
 	}
 }
 
+// maybeReleaseAffinity retrieves the pool and node for the given block and determines
+// if the pool still selects node. If it no longer matches, it will release the block
+// affinity for that node.
+// Returns a bool indicating if the block was released.
+func (c ipamClient) maybeReleaseAffinity(ctx context.Context, b *model.AllocationBlock) (bool, error) {
+	logCtx := log.WithField("cidr", b.CIDR)
+	// Retrieve node for this allocation. We do this so we can clean up affinity for blocks
+	// which should no longer be affine to this host.
+	host := getHostAffinity(b)
+	logCtx.Debugf("block affinity: %s", host)
+
+	// Exit function is host is empty.
+	if host == "" {
+		return false, nil
+	}
+
+	// If the IP pool which owns this block no longer selects this node,
+	// we should release the block's affinity to this node so it can be
+	// used elsewhere.
+	logCtx.Debugf("Looking up node for host affinity %s", host)
+	node, err := c.client.Get(ctx, model.ResourceKey{Kind: v3.KindNode, Name: host}, "")
+	if err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+			logCtx.WithError(err).WithField("node", host).Error("failed to get node for host")
+			return false, err
+		}
+		logCtx.Info("Node doesn't exist, releasing IPs anyway")
+		return false, nil
+	}
+
+	// Make sure the returned value is a valid node.
+	v3n, ok := node.Value.(*v3.Node)
+	if !ok {
+		return false, fmt.Errorf("Datastore returned malformed node object")
+	}
+
+	if pool := c.blockReaderWriter.getPoolForIP(cnet.IP{b.CIDR.IPNet.IP}); pool == nil {
+		// No IP pool owns this block, return.
+		return false, nil
+	} else if sel, err := pool.SelectsNode(*v3n); err != nil {
+		// Invalid selector syntax.
+		logCtx.WithError(err).WithField("selector", pool.Spec.NodeSelector).Error("failed to parse selector")
```

Comment:
nit: we don't know for sure this is an invalid selector. Should say "error checking if pool selects node"